File: android/src/main/kotlin/com/dra11y/flutter/native_segments/NativeSegmentsViewFactory.kt
package com.dra11y.flutter.native_segments

import android.content.Context
import io.flutter.plugin.common.JSONMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory
import org.json.JSONObject

class NativeSegmentsViewFactory(
    private val hostApi: NativeSegmentsHostApiAndroid,
    private val flutterApi: NativeSegmentsFlutterApi,
) : PlatformViewFactory(JSONMessageCodec.INSTANCE) {
    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        val id = (args as? JSONObject)?.optString("id") ?:
            throw Error("Instance ID must be passed from Flutter via JSONMessageCodec as { 'id': String }")
        return NativeSegmentsViewController(
            id = id,
            context = context,
            viewId = viewId,
            hostApi = hostApi,
            flutterApi = flutterApi,
        )
    }
}

File: ios/Classes/NativeSegmentsViewFactory.swift
//
//  NativeSegmentsViewFactory.swift
//  segments_ios
//
//  Created by Grushka, Tom on 9/22/22.
//

import Flutter
import Foundation

class NativeSegmentsViewFactory: NSObject, FlutterPlatformViewFactory {
    let hostApi: NativeSegmentsHostApiIOS
    let flutterApi: NativeSegmentsFlutterApi

    init(hostApi: NativeSegmentsHostApiIOS, flutterApi: NativeSegmentsFlutterApi) {
        self.hostApi = hostApi
        self.flutterApi = flutterApi
    }

    func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        FlutterJSONMessageCodec()
    }

    func create(withFrame frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?) -> FlutterPlatformView {
        guard
            let args = args as? [String : Any?],
            let id = args["id"] as? String
        else {
            preconditionFailure("Instance ID must be passed from Flutter via JSONMessageCodec as { 'id': String }")
        }
        return NativeSegmentsViewController(frame: frame, id: id, viewId: viewId, hostApi: hostApi, flutterApi: flutterApi)
    }
}
